### ISSUE TITLE
Add recurrent batch normalization for RNNs

### ISSUE THREAD

**Issue by lacic** [eclipse/deeplearning4j#2418]
BatchNormalization ( https://deeplearning4j.org/doc/org/deeplearning4j/nn/layers/normalization/BatchNormalization.html ) in DL4J can't currently be applied for RNN.

Is it possible to implement something like in the paper [Recurrent Batch Normalization](https://arxiv.org/abs/1603.09025) ? 
A ParameterSpace implementation for Arbiter would also be needed in that case.

**Comment by fac2003** [comment]
Layer Normalization should also be considered since it seems to work well for RNNs: https://arxiv.org/pdf/1607.06450.pdf

**Comment by raver119** [comment]
New batchnorm op available as CustomOp can be applied to time series

**Comment by maxpumperla** [comment]
sorry, but this is a few levels down from what's directly usable in dl4j. I can take this at some point.

**Comment by raver119** [comment]
If you want to have 10050 individual issues open - i’m ok with that. But
for me it’s the same issue - everything present in libnd4j should be mapped
to nd4j :)

**Comment by maxpumperla** [comment]
haha, if it makes you nervous I can close it again. We'll move to mono repo soon enough, but until then maybe we should treat libnd4j, nd4j and dl4j as the separate projects they are.

**Comment by raver119** [comment]
Huh, i'm just sorting out issues in favor of zenhub use.

**Comment by AlexDBlack** [comment]
@hmart027 Not yet.
We have batch norm implemented on the input (from previous layer) activations (that's just batch norm layer), but not on the recurrent activations. That would likely have to be built into the layer itself, which would be non-trivial to do.

**Comment by maxpumperla** [comment]
btw, this has also been requested in keras import land https://github.com/deeplearning4j/deeplearning4j/issues/4502, might be worth looking into at some point.

**Comment by reuschling** [comment]
Is there any further status on this? I try to implement a 1D CNN classification for timeseries from this [paper](https://arxiv.org/pdf/1611.06455.pdf) (the FCN, here is the according [Gist](https://gist.github.com/reuschling/ba771bb80b609b10cc3c328bdc489086)). They use batch normalization - I managed it to classify with 1D CNN layers now, but with BatchNormalization layers I get the 'Invalid input type' Exception with 'got InputTypeRecurrent(128,timeSeriesLength=23)'.

Maybe it is only about the `NeuralNetConfiguration=>listBuilder=>setInputType(InputType.recurrent(convNIn, length));`
and I could simply switch to a different InputType?
I've made a small [Gist](https://gist.github.com/reuschling/ba771bb80b609b10cc3c328bdc489086) with my code, thanks for answering!
